### PR TITLE
:sparkles: Improve image filtering

### LINKED ISSUE #116: Filter node-images in OpenStackNodeImageRelease controller based on ID or image name and tags
/kind feature

### Current state 

OpenStackNodeImageRelease controller filters images based on their names only. 

Background :

OpenStackNodeImageRelease resource name is composed as follows: `release tag without version`-`image name`-`node-image-version` (e.g.: `openstack-ferrol-1-27-ubuntu-capi-image-v1.27.8-v2`). When the OpenStackNodeImageRelease resource is created, the controller will search for the node image in the OpenStack project based on its `image name` only. If the OpenStack project does not contain a node image with the given image name, then the image is imported.

If the ClusterStack version is changed (e.g., the addon version has been updated), it is expected that the node image could be reused. This ensures that multiple versions of one ClusterStack can share one node image.

An issue could occur when the `node-image-version` is changed as well, e.g., some image create-options are changed, but the `image name` remains the same. In this case, the controller finds the given `image name` in the OpenStack project and does not import the new version of the image. This causes an inconsistent state because the node image in the OpenStack project was not created with the desired create-options and we end with one imported image and two OpenStackNodeImageRelease resources (two node-image-versions).  

### Desired state

OpenStackNodeImageRelease controller filters images based on their names and tags. 

We should allow the creation of multiple images with the same name in the OpenStack project and distinguish them by unique tags. Consequently, one ClusterStack, across multiple versions and importantly with multiple node-image-versions, could use the same image name. Therefore, the creator of the ClusterStack does not need to introduce a unique node-image name when the node-image-version is changed. However, the creator of the ClusterStack should ensure that the unique tag (node-image-version) is passed to the CAPO resources.

### CAPO 

The latest released version of Cluster API OpenStack Provider `v1alpha7` does not support image tag filtering.

In Cluster API OpenStack Provider `v1beta1` (not released yet) the node image could be filtered based on its ID or based on a combination of name, and tags (see [here](https://github.com/kubernetes-sigs/cluster-api-provider-openstack/blob/main/api/v1beta1/types.go#L25)). The combination of name and tags must return a single matching image. Filtering by tags is new and allows to have images with the same name, that should be recognized by tags.

### PR DESCRIPTION
**Pull request by michal-gubricky**
**What this PR does / why we need it**:
This PR improves image filtering

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #116 

**TODOs**:

- [ ] squash commits
- [ ] include documentation
- [ ] add unit tests

### CODE REVIEW

File: internal/controller/openstacknodeimagerelease_controller.go [hunk 1]
```diff
@@ -331,27 +331,35 @@ func (r *OpenStackNodeImageReleaseReconciler) getCloudFromSecret(ctx context.Con
 	return cloud, nil
 }
 
-func findImageByName(imagesClient *gophercloud.ServiceClient, imageName string) (string, error) {
-	listOpts := images.ListOpts{
-		Name: imageName,
+func getImageID(imagesClient *gophercloud.ServiceClient, image *apiv1alpha1.CreateOpts) (string, error) {
```
Comment (matofeder) [inline internal/controller/openstacknodeimagerelease_controller.go:334]: minor: rename `image` to `imageCreateOps`

File: internal/controller/openstacknodeimagerelease_controller.go [hunk 2]
```diff
@@ -331,27 +331,35 @@ func (r *OpenStackNodeImageReleaseReconciler) getCloudFromSecret(ctx context.Con
...
+	var listOpts images.ListOpts
+
+	if image.ID != "" {
+		return image.ID, nil
```
Comment (matofeder) [inline internal/controller/openstacknodeimagerelease_controller.go:338]: IMO we should not just return the image ID here if it is defined. 

We should check if the image with the given ID exists, and return the same ID if yes. 
If the image with the given ID does not exist we should return an empty string here, and then the image with the given ID will be crated later in code.

File: internal/controller/openstacknodeimagerelease_controller.go [hunk 3]
```diff
@@ -331,27 +331,35 @@ func (r *OpenStackNodeImageReleaseReconciler) getCloudFromSecret(ctx context.Con
...
-		if imageList[i].Name == imageName {
-			return imageList[i].ID, nil
-		}
+	switch len(imageList) {
+	case 0:
+		return "", nil
+	case 1:
+		return imageList[0].ID, nil
+	default:
+		return "", fmt.Errorf("too many images were found with the given image name: %s with tags: %s", imageCreateOps.Name, imageCreateOps.Tags)
```
Comment (matofeder) [inline internal/controller/openstacknodeimagerelease_controller.go:361]:
```suggestion
		return "", fmt.Errorf("too many images were found with the given image name: %s and tags: %s", imageCreateOps.Name, imageCreateOps.Tags)
```

File: internal/controller/openstacknodeimagerelease_controller.go (hunk 2 above)
Comment (michal-gubricky) [reply internal/controller/openstacknodeimagerelease_controller.go]: It depends on what kind of behavior you want. 

Do you want to create an image if the user/creator of the cluster-stacks enters the wrong image `ID`  in the `node-images. yaml`? In this scenario he does not need to set the `URL`, does he?

**Review by matofeder — APPROVED** [review]
lgtm

**Pull request merged**